File: packages/revocation/go/inmem.go
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// In-memory Store with per-entry TTLs.

package revocation

import (
	"context"
	"sync"
	"time"
)

// InMemoryStore is a process-local Store backed by a TTL map.
type InMemoryStore struct {
	mu      sync.Mutex
	entries map[string]time.Time
	defTTL  time.Duration
}

// NewInMemoryStore returns a Store using defaultTTL for entries that omit a TTL.
func NewInMemoryStore(defaultTTL time.Duration) *InMemoryStore {
	if defaultTTL <= 0 {
		defaultTTL = 24 * time.Hour
	}
	return &InMemoryStore{entries: map[string]time.Time{}, defTTL: defaultTTL}
}

// IsRevoked reports whether sid is currently revoked, evicting expired entries.
func (s *InMemoryStore) IsRevoked(_ context.Context, sid string) bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	expiresAt, ok := s.entries[sid]
	if !ok {
		return false
	}
	if !time.Now().Before(expiresAt) {
		delete(s.entries, sid)
		return false
	}
	return true
}

// MarkRevoked records sid as revoked for ttl, falling back to the default TTL when zero.
func (s *InMemoryStore) MarkRevoked(sid string, ttl time.Duration) {
	if ttl <= 0 {
		ttl = s.defTTL
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	s.entries[sid] = time.Now().Add(ttl)
}

File: packages/revocation/ts/src/inmem.ts
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// In-memory RevocationStore with per-entry TTLs and bounded capacity.

import type { RevocationStore } from './iface.js'

interface Entry {
  expiresAt: number
}

export class InMemoryRevocationStore implements RevocationStore {
  private readonly entries = new Map<string, Entry>()
  private readonly defaultTtlMs: number
  private readonly maxEntries: number

  constructor(opts: { defaultTtlMs?: number; maxEntries?: number } = {}) {
    this.defaultTtlMs = opts.defaultTtlMs ?? 24 * 60 * 60 * 1000
    this.maxEntries = opts.maxEntries ?? 100_000
  }

  isRevoked(sid: string): boolean {
    const entry = this.entries.get(sid)
    if (!entry) return false
    if (entry.expiresAt <= Date.now()) {
      this.entries.delete(sid)
      return false
    }
    return true
  }

  markRevoked(sid: string, ttlMs?: number): void {
    if (this.entries.size >= this.maxEntries) {
      this.reapExpired()
      if (this.entries.size >= this.maxEntries) throw new Error('Revocation store capacity exceeded')
    }
    this.entries.set(sid, { expiresAt: Date.now() + (ttlMs ?? this.defaultTtlMs) })
  }

  size(): number {
    return this.entries.size
  }

  private reapExpired(): void {
    const now = Date.now()
    for (const [sid, entry] of this.entries) {
      if (entry.expiresAt <= now) this.entries.delete(sid)
    }
  }

}
